File: map2d/src/main/java/net/ponderingprogrammer/map2d/ChebyshevFixedSquareMap2d.java
package net.ponderingprogrammer.map2d;

import java.util.ArrayList;
import java.util.Collection;

public class ChebyshevFixedSquareMap2d<T> extends FixedSquareMap2d<T> {
    public ChebyshevFixedSquareMap2d(int width, int height) {
        super(width, height);
    }

    @Override
    public Collection<Cell<T>> findAdjacentCells(int x, int y) {
        ArrayList<Cell<T>> adjacents = new ArrayList<>();
        getCell(x,y - 1).ifPresent(adjacents::add);    // top
        getCell(x + 1,y - 1).ifPresent(adjacents::add);    // top-right
        getCell(x + 1,y).ifPresent(adjacents::add);     // right
        getCell(x + 1,y + 1).ifPresent(adjacents::add);     // bottom-right
        getCell(x,y + 1).ifPresent(adjacents::add);     // bottom
        getCell(x - 1, y + 1).ifPresent(adjacents::add);   // bottom-left
        getCell(x - 1,y).ifPresent(adjacents::add);    // left
        getCell(x - 1,y - 1).ifPresent(adjacents::add);   // top-left
        return adjacents;
    }

    @Override
    public CellGraph<T> toCellGraph() {
        var graph = new CellGraph<T>();
        for (Cell<T> cell: getCells()){
            graph.getNodes().add(cell);
            for (Cell<T> adjacent: findAdjacentCells(cell.getCoord())) {
                if (cell == adjacent) continue;
                graph.getEdges().add(new CellGraphEdge<>(cell, adjacent));
            }
        }
        return graph;
    }
}

File: projects/dotnet/NProcGen/src/PonderingProgrammer.Map2d/ChebyshevFixedSquareMap2d.cs
using System.Collections.Generic;

namespace PonderingProgrammer.Map2d
{
    public class ChebyshevFixedSquareMap2d<T> : FixedSquareMap2d<T>
    {
        public ChebyshevFixedSquareMap2d(int width, int height) : base(width, height)
        {
        }

        public override IEnumerable<Cell<T>> FindAdjacentCells(int x, int y)
        {
            if (GetCell(x, y - 1) is { } top) yield return top;
            if (GetCell(x + 1, y - 1) is { } topRight) yield return topRight;
            if (GetCell(x + 1, y) is { } right) yield return right;
            if (GetCell(x + 1, y + 1) is { } bottomRight) yield return bottomRight;
            if (GetCell(x, y + 1) is { } bottom) yield return bottom;
            if (GetCell(x - 1, y + 1) is { } bottomLeft) yield return bottomLeft;
            if (GetCell(x - 1, y) is { } left) yield return left;
            if (GetCell(x - 1, y - 1) is { } topLeft) yield return topLeft;
        }

        public override CellGraph<T> ToCellGraph()
        {
            var graph = new CellGraph<T>();
            foreach(var cell in GetCells())
            {
                graph.Nodes.Add(cell);
                foreach(var adjacent in FindAdjacentCells(cell.GridCoord))
                {
                    if (cell == adjacent) continue;
                    graph.Edges.Add(new CellGraphEdge<T>(cell, adjacent));
                }
            }
            return graph;
        }
    }
}
